src/modules/roc_netio/target_uv/roc_netio/transceiver.cpp

```diff
@@ -214,76 +200,63 @@ void Transceiver::stop_sem_cb_(uv_async_t* handle) {
     roc_panic_if_not(handle);
 
     Transceiver& self = *(Transceiver*)handle->data;
-    self.stop_();
-    self.close_();
+    self.async_close_ports_();
+    self.close_sems_();
     self.process_tasks_();
 }
 
-void Transceiver::stop_() {
-    for (core::SharedPtr<UDPReceiver> rp = receivers_.front(); rp;
-         rp = receivers_.nextof(*rp)) {
-        rp->stop();
+void Transceiver::close_sems_cb_(uv_handle_t* handle) {
+    roc_panic_if_not(handle);
+
+    Transceiver& self = *(Transceiver*)handle->data;
+
+    core::Mutex::Lock lock(self.mutex_);
+
+    if (handle == (uv_handle_t*)&self.stop_sem_) {
+        self.stop_sem_initialized_ = false;
+    } else {
+        self.task_sem_initialized_ = false;
     }
 
-    for (core::SharedPtr<UDPSender> sp = senders_.front(); sp;
-         sp = senders_.nextof(*sp)) {
-        sp->stop();
+    if (self.stop_sem_initialized_ || self.task_sem_initialized_) {
+        return;
     }
+
+    self.cond_.broadcast();
 }
 
-void Transceiver::close_() {
-    if (task_sem_initialized_) {
-        uv_close((uv_handle_t*)&task_sem_, NULL);
-        task_sem_initialized_ = false;
-    }
+void Transceiver::async_close_ports_() {
```

Comment:
async_close_ports_() should acquire a mutex because it modifies the lists.

Also, now, when it removes ports from open_ports_, it can be simplified a bit:

```
    while (core::SharedPtr<BasicPort> port = open_ports_.front()) {
        open_ports_.remove(*port);
        closing_ports_.push_back(*port);

        port->async_close();
    }
```

Reply:
Done